### PR TITLE
[SC-641] CCTP split transfers

### PR DESCRIPTION
**Pull request by hexonaut**
Will split the CCTP transfers automatically into separate pieces if larger than the burn limit.

### CODE REVIEW

File: src/ForeignController.sol [hunk 1]
```diff
@@ -117,27 +140,48 @@ contract ForeignController is AccessControl {
...
+                abi.encodeCall(
+                    cctp.depositForBurn,
+                    (
+                        usdcAmount,
+                        destinationDomain,
+                        mintRecipient,
+                        address(usdc)
+                    )
+                )
+            );
```
Comment (barrutko) [inline src/ForeignController.sol:175]: I'd extract it to sth like `function initiateCCTPTransfer(uint256) internal`

File: src/ForeignController.sol [hunk 2]
```diff
@@ -16,11 +19,14 @@ interface ICCTPLike {
         bytes32 mintRecipient,
         address burnToken
     ) external returns (uint64 nonce);
+    function localMinter() external view returns (ICCTPTokenMinterLike);
 }
 
-contract ForeignController is AccessControl {
+interface ICCTPTokenMinterLike {
+    function burnLimitsPerMessage(address) external view returns (uint256);
+}
```
Comment (barrutko) [inline src/ForeignController.sol:27]: I'd move `ICCTPTokenMinterLike` and `ICCTPLike` to separate files

File: src/ForeignController.sol [hunk 3]
```diff
@@ -117,27 +140,48 @@ contract ForeignController is AccessControl {
...
-            abi.encodeCall(
-                cctp.depositForBurn,
-                (
-                    usdcAmount,
-                    destinationDomain,
-                    mintRecipient,
-                    address(usdc)
+        // If amount is larger than limit we must break it up
+        uint256 burnLimit = cctp.localMinter().burnLimitsPerMessage(address(usdc));
+        while (usdcAmount > burnLimit) {
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/ForeignController.sol:145]: There's not really any danger of running out of gas. Even at the block limit the worst that happens is the transaction won't submit. Maybe it's worthwhile to add a test to see what the gas limits are, but I would consider this the responsibility of the planner to not try to submit some really high number in the same way it respects the rate limits.

File: src/ForeignController.sol (hunk 2 above)
Comment (hexonaut) [reply src/ForeignController.sol:27]: I did this, but I made another task to just use the project interfaces: https://github.com/circlefin/evm-cctp-contracts/tree/master/src/interfaces

File: test/mainnet-fork/CCTPCalls.t.sol [hunk 4]
```diff
@@ -286,4 +357,72 @@ contract USDCToCCTPIntegrationTests is BaseChainUSDCToCCTPTestBase {
...
+        );
+        vm.expectEmit(CCTP_MESSENGER_BASE);
+        emit DepositForBurn(
+            255143,
+            address(usdcBase),
+            600_000e6,
+            address(foreignAlmProxy),
+            foreignController.mintRecipients(CCTPForwarder.DOMAIN_ID_CIRCLE_ETHEREUM),
+            CCTPForwarder.DOMAIN_ID_CIRCLE_ETHEREUM,
+            bytes32(0x000000000000000000000000bd3fa81b58ba92a82136038b25adec7066af3155),
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/mainnet-fork/CCTPCalls.t.sol]: I moved to function.

File: src/ForeignController.sol (hunk 1 above)
Comment (hexonaut) [reply src/ForeignController.sol]: Good call. Done.

File: src/ForeignController.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (barrutko) [reply src/ForeignController.sol:133]: Ok, yeah, maybe it makes sense. On different domains there can be different block sizes, maybe it's the planner who should be aware of all these limitations and come up with a number that will make sense. Let's keep it as is.

File: test/mainnet-fork/CCTPCalls.t.sol [hunk 5]
```diff
@@ -255,6 +246,47 @@ contract USDCToCCTPIntegrationTests is BaseChainUSDCToCCTPTestBase {
...
+        assertEq(usdcBase.balanceOf(address(foreignAlmProxy)),   0);
+        assertEq(usdcBase.balanceOf(address(foreignController)), 0);
+        assertEq(usdcBase.totalSupply(),                         USDC_BASE_SUPPLY);
+
+        bridge.relayMessagesToDestination(true);
+
+        assertEq(usdcBase.balanceOf(address(foreignAlmProxy)),   1_900_000e6);
+        assertEq(usdcBase.balanceOf(address(foreignController)), 0);
+        assertEq(usdcBase.totalSupply(),                         USDC_BASE_SUPPLY + 1_900_000e6);
+    }
```
Comment (lucas-manuel) [inline test/mainnet-fork/CCTPCalls.t.sol:288]: Update to do multiple iterations of while loop:


```suggestion
    function test_transferUSDCToCCTP_sourceToDestination_bigTransfer() external {
        deal(address(usdc), address(almProxy), 2_900_000e6);

        assertEq(usdc.balanceOf(address(almProxy)),          2_900_000e6);
        assertEq(usdc.balanceOf(address(mainnetController)), 0);
        assertEq(usdc.totalSupply(),                         USDC_SUPPLY);

        assertEq(usds.allowance(address(almProxy), CCTP_MESSENGER),  0);

        // Will split into two separate transactions at max 1m each
        _expectEthereumCCTPEmit(
            94773,
            1_000_000e6
        );
        _expectEthereumCCTPEmit(
            94774,
            1_000_000e6
        );
        _expectEthereumCCTPEmit(
            94775,
            900_000e6
        );

        vm.prank(relayer);
        mainnetController.transferUSDCToCCTP(2_900_000e6, CCTPForwarder.DOMAIN_ID_CIRCLE_BASE);

        assertEq(usdc.balanceOf(address(almProxy)),          0);
        assertEq(usdc.balanceOf(address(mainnetController)), 0);
        assertEq(usdc.totalSupply(),                         USDC_SUPPLY - 2_900_000e6);

        assertEq(usds.allowance(address(almProxy), CCTP_MESSENGER),  0);

        destination.selectFork();

        assertEq(usdcBase.balanceOf(address(foreignAlmProxy)),   0);
        assertEq(usdcBase.balanceOf(address(foreignController)), 0);
        assertEq(usdcBase.totalSupply(),                         USDC_BASE_SUPPLY);

        bridge.relayMessagesToDestination(true);

        assertEq(usdcBase.balanceOf(address(foreignAlmProxy)),   2_900_000e6);
        assertEq(usdcBase.balanceOf(address(foreignController)), 0);
        assertEq(usdcBase.totalSupply(),                         USDC_BASE_SUPPLY + 2_900_000e6);
    }
```

Comment (hexonaut) [reply test/mainnet-fork/CCTPCalls.t.sol:288]: updated

**Review by lucas-manuel — APPROVED** [review]

File: test/mainnet-fork/CCTPCalls.t.sol [hunk 6]
```diff
@@ -302,4 +325,83 @@ contract USDCToCCTPIntegrationTests is BaseChainUSDCToCCTPTestBase {
...
+        assertEq(usdcBase.balanceOf(address(foreignController)), 0);
+        assertEq(usdcBase.totalSupply(),                         USDC_BASE_SUPPLY);
+
+        assertEq(usdsBase.allowance(address(foreignAlmProxy), CCTP_MESSENGER_BASE),  0);
+
+        // Will split into three separate transactions at max 1m each
+        _expectBaseCCTPEmit(
+            255141,
+            1_000_000e6
+        );
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/mainnet-fork/CCTPCalls.t.sol:347]: Oops I missed this, but let's just get this in for now.

**Pull request merged**